src/providers/json-rpc-provider.ts

```diff
@@ -56,9 +56,14 @@ export class JsonRpcProvider extends Provider {
     /**
      * @param url RPC API endpoint URL
      */
-    constructor(url?: string) {
+    constructor(connectionInfoOrUrl?: string | ConnectionInfo) {
         super();
-        this.connection = { url };
+        if (!connectionInfoOrUrl) { return; }
```

Comment:
This represents a behavioural change that would need to be classified as a breaking change, given that consumers can reference `JsonRpcProvider.connection` directly.

Previously, `this.connection` would **always** at least be an object, even if it only had 1 property called `url` that might be `undefined` or an empty string.  

Now, `this.connection` will be `undefined` in this case (`url` is undefined or empty string), which could represent a breaking change for anyone who references `JsonRpcProvider.connection.xxxx` in their code when they have constructed the provider with an empty string for URL, or no url -- they would start suddenly getting errors thrown `cannot read xxx of undefined`.

Reply:
Fixed, test added. Thanks!

Reply:
1. We can't use the same logic here as we do in `fetchJson()`, because  connectionInfoOrUrl is optional.  To ensure that this behaviour remains the same for anyone calling this method without an URL, we would actually need to invert this check and only use the new logic if passed an **object**...

2. The check on L62 is a weak equality check which will coerce falsy values, one of which is an empty string.  If we were given an empty string previously, then that empty string would become `this.connection.url` -- but with the early exit on L62, now `this.connection.url` will be `null` which will break any existing strict equality checks.

I think this would retain existing behaviour while adding support for `url` to be a `ConnectionInfo` object.
```javascript
    constructor(connectionInfoOrUrl?: string | ConnectionInfo) {
        super();

        if (value != null && typeof value == 'object') {
            this.connection = connectionInfoOrUrl as ConnectionInfo;
        } else {
            this.connection = { url: connectionInfoOrUrl };            
        }
    }
```
What do you think?

3. Should we mark the previous function signature as deprecated like we did with other functions that we migrated from positional arguments to taking a single configuration object?

4. JSDoc for this function needs to be updated or our generated near-api-js API docs will be incorrect.

Reply:
1, 2, 4 -> fixed/added. Thanks :)
3. It's a constructor, so I just added a deprecation message.